doc/LANGUAGE.md

```diff
@@ -345,26 +345,36 @@ Instruction                 | Flags   | Description
 
 The `const` instruction is used to introduce a constant value into the IR. The first version constructs a constant integer value, the second a constant integer signal, and the third a constant time value.
 
-    %result = const iN <int>
-    %result = const iN$ <int>
     %result = const time <time>
+    %result = const iN <int>
+    %result = const nN <enum>
+    %result = const lN <logic>
 
+- `time` is a time literal such as `1s`, `1s 2d`, or `1s 2d 3e`, where the real component may carry an SI suffix such as `as`, `fs`, `ps`, `ns`, `us`, `ms`, `s`.
 - `int` is an integer literal such as `0b0101`, `0o1247`, `129`, or `0x14F3E`
-- `time` is a time literal such as `1s`, `1s 2d`, or `1s 2d 3e`, where the real component may carry an SI suffix such as `as`, `fs`, `ps`, `ns`, `us`, `ms`.
+- `enum` is a decimal integer literal which ranges from `0` to `N-1`
+- `logic` is a string of `N` logic value characters (one of `U`, `X`, `0`, `1`, `Z`, `W`, `L`, `H`, `-`)
 
 ##### Example
 
+A constant time with value 1s+3d+7e may be constructed as follows:
+
+    %0 = const time 1s 3d 7e
+
 A constant 32 bit integer with value 42 may be constructed as follows:
 
     %0 = const i32 42
-    %1 = const i32$ 42
     ; type(%0) = i32
-    ; type(%1) = i32$
 
-A constant time with value 1s+3d+7e may be constructed as follows:
+A constant integer ranging from `0` to `99` with value 13 may be constructed as follows:
 
-    %0 = const time 1s 3d 7e
+    %0 = const n100 13
+    ; type(%0) = n100
+
+A constant value for 4 logic wires may be constructed as follows (note: here, the _lowest_ indexed wire is a logic `Z` and the _highest_ indexed wire is a logic `L`):
 
+    %0 = const l4 "L0LZ"
+    ; type(%0) = l4
```

Comment:
We could inherit Verilog's handling of logic values, which is pretty sane (and an improvement over VHDL): Instead of a string, the user can just write any of the binary, octal, or hexadecimal numbers (not decimal though), but with an extended digit set (the ones you listed above). A corresponding number of bits will then be assigned the logic value.

For example:
- `0bL0LZ` gives you the existing example `"L0LZ"`
- `0oLOLZ` gives you `"LLLOOOLLLZZZ"`
- `0xLOLZ` gives you `"LLLLOOOOLLLLZZZZ"

Reply:
This also sounds good to me :) It would mean that the string syntax no longer exists, right? If so, this is a change that would "break existing code" (I suppose this is fine since you were saying that examples needed updating... But I think it is the first suggestion that I see that led to something like this so might as well raise a flag there just to be sure...).

Reply:
Having another think about this (as it is in fact the first time I am made aware of this Verilog feature) ...

If I understand correctly, this means that `0x01` would mean the integer `1` when used to construct a `i8` or a `n8`, but would effectively map to integer `15` when used to construct a `l8`. Is that the case?

I am not sure if I judge this to be awkward or not, but it certainly sounds note-worthy. I presume since this is a Verilog feature, it is generally accepted syntax and people should "just be aware"?

Reply:
> This also sounds good to me :) It would mean that the string syntax no longer exists, right? If so, this is a change that would "break existing code" (I suppose this is fine since you were saying that examples needed updating... But I think it is the first suggestion that I see that led to something like this so might as well raise a flag there just to be sure...).

Yes exactly, there wouldn't  be any string syntax. I think there is no code currently accepted by the parser where string syntax is used, so it doesn't break existing code. In fact, adding proper `lN` handling has been on the todo list for a while. But it's nice that the spec gets cleaned before the impl is hacked together :wink:.

Reply:
> If I understand correctly, this means that `0x01` would mean the integer `1` when used to construct a `i8` or a `n8`, but would effectively map to integer `15` when used to construct a `l8`. Is that the case?

The numeric digits work the same as you would expect. So `0x01` still is `00000001`. Instead of numeric digits you can now also use e.g. `X` or `Z`, which sets all bits that would be occupied by that digit to `X` or `Z`. So `0xZ1` would be `ZZZZ0001`, and `0oZ1` would be `ZZZ001`.

Reply:
> The numeric digits work the same as you would expect. So `0x01` still is `00000001`. Instead of numeric digits you can now also use e.g. `X` or `Z`, which sets all bits that would be occupied by that digit to `X` or `Z`. So `0xZ1` would be `ZZZZ0001`, and `0oZ1` would be `ZZZ001`.

I see... So, since there is no overlap in letters A-F and the ones used in the definition of a single logic value, can we use a full range of hex values there? As in `0xDZ` would be `1101ZZZZ`, or also `0o6-X` for `110---XXX`? Or is it limited to `0` and `1` exclusively?

Reply:
Yes exactly, you have the full range of digits available. It's pretty much the regular number literal, but in addition to `[0-9A-F]`, you can use e.g. `X` to replace the 1/3/4 bits occupied by a digit with `X`.

Reply:
I see. That's nice.